### ISSUE TITLE
RFC: `hydrate.copy` API

### ISSUE THREAD

**Issue by ryanblock** [architect/architect#1369]
This issue is for collecting feedback on the `hydrate.copy` API, currently experimental as of Architect 10.5.

The goal of the `hydrate.copy` API is to give Architect plugin developers a hook for copying arbitrary files and folders into all Lambdas after dependency and shared code operations are complete, thus ensuring that such filesystem mutations will be present in deployment.

For more information about the API, please refer to the docs. Would love to hear what you think here about `hydrate.copy` (and other `hydrate.*` API ideas) here!

**Comment by tbeseda** [comment]
Hey @ryanblock is it required that an Arc app have shared enabled in order for a `hydrate.copy` plugin?

Right now, if an Arc project doesn't have src/shared/ (or `@shared` set up in the manifest), the Sandbox will throw when a plugin provides `hydrate.copy`

Here's [a sample repo with 2 commits](https://github.com/tbeseda/scratch-arc-hydrate-copy-no-shared/commits/main); the first demonstrates the broken state and the second shows it works when src/shared/ is added.

The error from the broken state:
```
[...]/arc-hydrate-copy-no-shared/node_modules/@architect/hydrate/src/shared/copy-plugins.js:14
  if (copyPlugins && shared.length) {
                            ^

TypeError: Cannot read properties of undefined (reading 'length')
```

I took a look at the code here and `shared` (from `inventory.inv`) is used inside the `copy` function sent to the plugin invocation that provides `hydrate.copy`. So I wasn't able to fully grok how we might do this without `shared` in that scope.

**Comment by tbeseda** [comment]
Yeah I see the potential trouble with hydrating something to all Lambdas. Maybe `lambdaA` is a big ol' function that barely fits in the allotted space but I want to hydrate something to `lambdaB` and `lambdaC`. That hydration could put `lambdaA` over the edge.

What if the config to exclude a Lambda a part of the plugin instead of in the consumer's .arc?

The provided `copy` function could take objects with an optional third attribute, like an evaluator function that returns a boolean to decide if a target should receive a copy of the `src`

```js
module.exports = {
  hydrate: {
    async copy ({ copy }) {
      await copy({
        source: "project/relative/path/file.txt",
        target: "file.txt",
        doHydrate(target) {
          return target.path.indexOf('not-lambdaA') < 0;
        }
      })
    },
  },
}
```

A plugin could do a few cool things like require its own config or even analyze that target Lambda to decide if it needs that file.

I'm not married to it, just an idea :)

**Comment by ryanblock** [comment]
Hmm, clever idea. I had also been thinking of the opposite: ONLY share to a given Lambda.

I'm thinking that at least just to start (and since this API is still in beta) how about we change it to hydrate all Lambdas, regardless of shared code. And if folks want this to be more filtered we can hear that feedback and add filtering features.

Here's a PR, lmk what you think: https://github.com/architect/hydrate/pull/163

**Comment by snelgrove** [comment]
If you're still gathering feedback, I'm working on a plugin which has raised this scenario:
- My plugin creates several routes
- These routes share common code which is only intended for internal plugin use
- `hydrate.copy` shares this common code amongst all routes in the application

Ideally, I'd be able to restrict `hydrate.copy` to internal plugin routes, and not expose plugin specific code to the consuming app. 

A generic workaround might be the option to specify a regex filter for the target path. For example, today I can specify a target of `@architect/shared` to copy something into the shared directory of *every* Lambda listed in `lambdaSrcDirs`. Setting a filter that applies to `lambdaSrcDirs` would allow finer control:
- A regex *including* `node_modules/<my-plugin-name>` would skip non-plugin routes
- A regex *excluding* `node_modules/<my-plugin-name>` would skip plugin routes and copy only to the consuming app
- Developers can write even more specific filters, and even call copy multiple times with different source, target and route filters

For example: 
```js
hydrate: {
  copy: async ( { arc, inventory, copy } ) => {
    await copy( { 
      source: 'node_modules/my-plugin/src/shared',
      target: 'my-plugin/',
      targetFilter: /^((?!(node_modules\/my-plugin)).)*$/ // Exclude plugin routes
    })
  }
}
```
